### PR TITLE
Canvas: refactor layer editor

### LINKED ISSUE #42095: Canvas: Refactor Layer Editor to be Reusable Between Canvas / Geomap Panel
Refactor layer editor to be fully reusable between canvas and geomap panels

### PR DESCRIPTION
**Pull request by nmarrs**
**What this PR does / why we need it**:
Refactor layer editor to be fully reusable between canvas and geomap panels. This should allow us to make improvements to layer editor and directly benefit both canvas and geomap panels.

Add ability to edit element names in canvas panel:
Screen Shot 2021-12-01 at 12 38 00 PM

**Which issue(s) this PR fixes**:

Fixes #42095

### CODE REVIEW

File: public/app/plugins/panel/geomap/editor/LayerDragDropList/LayerDragDropList.tsx [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                  <Draggable key={uid} draggableId={uid} index={rows.length}>
+                    {(provided, snapshot) => (
+                      <div
+                        className={getRowStyle(isSelected)}
+                        ref={provided.innerRef}
+                        {...provided.draggableProps}
+                        {...provided.dragHandleProps}
+                        onMouseDown={() => onSelect(element)}
+                      >
+                        <LayerName layer={element} verifyLayerNameUniqueness={verifyLayerNameUniqueness ?? undefined} />
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/editor/LayerDragDropList/LayerDragDropList.tsx:66]: looks like the `element` (T) should b an interface with a "name" and "type"?

File: public/app/plugins/panel/geomap/editor/LayerDragDropList/LayerDragDropList.tsx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                        <div className={style.textWrapper}>&nbsp; {element.options.type}</div>
```
Comment (ryantxu) [inline public/app/plugins/panel/geomap/editor/LayerDragDropList/LayerDragDropList.tsx:67]:
```suggestion
                        <div className={style.textWrapper}>&nbsp; {props.getInfo(element)}</div>
```
Maybe a callback function to get the text?

File: public/app/plugins/panel/geomap/editor/LayerDragDropList/LayerDragDropList.tsx (hunk 1 above)
Comment (nmarrs) [reply public/app/plugins/panel/geomap/editor/LayerDragDropList/LayerDragDropList.tsx:72]: I may need a bit of help on getting types nailed down, but other than that I think PR is in pretty good shape

File: public/app/plugins/panel/canvas/editor/elementEditor.tsx [hunk 3]
```diff
@@ -68,6 +68,7 @@ export function getElementEditor(opts: CanvasEditorOptions): NestedPanelOptions<
         currentOptions = {
           ...layer.getNewOptions(options),
           type: layer.id,
+          name: `Element ${Date.now()}.${Math.floor(Math.random() * 100)}`,
```
Comment (ryantxu) [inline public/app/plugins/panel/canvas/editor/elementEditor.tsx:71]: In a new PR, lets resolve getting a unique name from the scene -- and having the scene keep a name > element Map (currently based on UID)

**Review by ryantxu — APPROVED** [review]

**Pull request merged**